### PR TITLE
Fabric events Lifecycle v0

### PR DESCRIPTION
**Pull request by BoogieMonster1O1**
Adds lots of useful events. 
and applies the license header to all java files

Client Events:-
* `ClientStartCallback` - Runs when the client starts
* `ClientStopCallback` - Runs when the client is scheduled to stop
* `ClientTickCallback` - Runs every tick on the client
* `ItemTooltipCallback` - Runs when an item tooltip is built
* `LanServerPublishedCallback` - Runs when the integrated server is published

Server Events:-
* `ServerStartCallback` - Runs when the server starts
* `ServerStopCallback` - Runs before the server shuts down
* `ServerTickCallback` - Runs every tick on the server
* `PlayerJoinCallback` - Runs every time a player joins a world
* `DedicatedServerSetupCallback` - Runs after the dedicated server is set up, which is shortly after starting

World Events:-
* `ChunksSavedCallback` - Runs when all chunks are successfully saved
* `EntityHurtCallback` - Runs when an entity is hurt
* `EntityKilledCallback` - Runs when an entity is killed
* `ServerPlayerTickCallback`- Runs every tick on a `ServerPlayerEntity`
* `WorldTickCallback` - Runs every tick on a `World`

### CODE REVIEW

File: fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/mixin/content/registries/WorldMixin.java [hunk 1]
```diff
@@ -103,4 +87,21 @@ public void init(class_635 arg, LevelProperties levelProperties, Dimension dimen
 			}
 		}
 	}
+
+	private void writeIdsToFile(BiMap<Integer, Identifier> idMap, File blockIds) throws IOException {
```
Comment (shedaniel) [inline fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/mixin/content/registries/WorldMixin.java:91]: Add @Unique

Comment (shedaniel) [reply fabric-content-registries-v1/src/main/java/net/fabricmc/fabric/mixin/content/registries/WorldMixin.java:91]: change `blockIds` to `file`

File: fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/api/event/world/EntityHurtCallback.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import net.minecraft.entity.damage.DamageSource;
+
+public interface EntityHurtCallback {
+	Event<EntityHurtCallback> EVENT = EventFactory.createArrayBacked(EntityHurtCallback.class, (listeners) -> (entity,source,original,damage) -> {
+		for (EntityHurtCallback callback : listeners) {
+			callback.chunksSaved(entity,source,original,damage);
+		}
+	});
+
+	void chunksSaved(LivingEntity entity, DamageSource damageSource, float originalHealth, float damage);
```
Comment (shedaniel) [inline fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/api/event/world/EntityHurtCallback.java:31]: wrong method name

File: fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/mixin/event/lifecycle/MixinMinecraftClient.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	@Inject(at = @At("HEAD"),method = "scheduleStop")
+	public void stop(CallbackInfo info){
+		ClientStopCallback.EVENT.invoker().onStopClient((MinecraftClient) (Object) this);
+	}
+
+	@Inject(at = @At("HEAD"),method = "run")
+	public void start(CallbackInfo info){
+		ClientStopCallback.EVENT.invoker().onStopClient((MinecraftClient) (Object) this);
```
Comment (shedaniel) [inline fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/mixin/event/lifecycle/MixinMinecraftClient.java:44]: Is this correct? why are we calling stop on start?

File: fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/mixin/event/lifecycle/MixinPlayerManager.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(PlayerManager.class)
+public class MixinPlayerManager {
+	@Inject(at=@At("HEAD"),method = "onPlayerConnect")
+	public void onPlayerConnect(ClientConnection connection, ServerPlayerEntity player, CallbackInfo info){
+		PlayerJoinCallback.EVENT.invoker().playerJoin(connection,player);
```
Comment (shedaniel) [inline fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/mixin/event/lifecycle/MixinPlayerManager.java:32]: PlayerJoinCallback -> PlayerConnectCallback

File: fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/MixinKeyedResourceReloadListener.java [hunk 5]
```diff
@@ -39,7 +39,7 @@
 
 public class MixinKeyedResourceReloadListener {
 	@Mixin({
-			SoundManager.class, GameRenderer.class, LanguageManager.class, class_1255.class, class_1254.class, TextureManager.class,
+			SoundManager.class, EntityRenderer.class, LanguageManager.class, class_1255.class, class_1254.class, TextureManager.class,
```
Comment (shedaniel) [inline fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/mixin/resource/loader/MixinKeyedResourceReloadListener.java:42]: concern when did this get renamed.

File: fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/NbtType.java [hunk 6]
```diff
@@ -17,15 +17,14 @@
...
 
 import net.minecraft.nbt.CompoundTag;
-import net.minecraft.nbt.Tag;
+import net.minecraft.nbt.NbtTag;
 
 /**
  * NBT type ID constants. Useful for filtering by value type in a few cases.
  *
- * <p>For the current list of types, check with {@link Tag#TYPES}.
+ * <p>For the current list of types, check with {@link NbtTag#TYPES}.
```
Comment (shedaniel) [inline fabric-api-base/src/main/java/net/fabricmc/fabric/api/util/NbtType.java:25]: concern when did this get renamed

File: fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/api/event/world/EntityHurtCallback.java (hunk 2 above)
Comment (BoogieMonster1O1) [reply fabric-events-lifecycle-v0/src/main/java/net/fabricmc/fabric/api/event/world/EntityHurtCallback.java:31]: Fixed with `onChunksSaved`

**Comment by BoogieMonster1O1** [comment]
New Registry Events:-
`RegistryBlockAddedCallback` - Runs every time a block is registered in `BlockRegistry`
`RegistryItemAddedCallback` - Runs every time an item or blockitem is registered in `ItemRegistry`

**Comment by liach** [comment]
yeah looks good, can always run formatter later

**Comment by liach** [comment]
> 
> 
> This all looks good. by the way, dont request reviews from liach if its not mappings. i dont think they want random pr's to review

oh i just went on something else halfway through review, feel free to request

**Comment by hYdos** [comment]
oh ok good to know

**Pull request merged**